### PR TITLE
Feature/22/cli init

### PR DESCRIPTION
**Pull request by cliffle**
This PR introduces the command `sakuli create project` concerning the issue #22.

### CODE REVIEW

File: packages/sakuli-legacy/src/command/create-commands/create-project-command.class.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+            }).demandOption('directory')
+            .positional('suiteName', {
+                describe: 'name of the testsuite',
+                default: 'sakuli'
+            });
+    },
+    async handler(opts: any) {
+        console.log("creating project structure");
+        createTestsuite(opts.directory, opts.suiteName);
+        process.exit(0);
```
Comment (tnobody) [inline packages/sakuli-legacy/src/command/create-commands/create-project-command.class.ts]: We should provide more details to the user here:
- `creating project in $absolutePathInUserFileSystem`
- `successfully create testsuite`
and so on.

Additionally we should check if `opts.directory` is a sakuli project root otherwise we should warn the user. We can also think about creating a proper package.json file and install sakuli there.

I actually think that the directory can also be optional with a default at the current workdirectory where the user runs the command (`process.pwd()`).

What do you think @s1hofmann  @svettwer

Comment (ZNicoZ) [reply packages/sakuli-legacy/src/command/create-commands/create-project-command.class.ts]: test 2

File: packages/sakuli-legacy/src/init-command/create-structure.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        },
+        author: "",
+        license: "ISC",
+        dependencies: {
+            "@sakuli/cli": "^2.1.3",
+        },
+    };
+
+    writeFileSync(`${path}/package.json`, JSON.stringify(packageJson, null, 4));
+    console.log("Created file package.json")
```
Comment (tnobody) [inline packages/sakuli-legacy/src/init-command/create-structure.ts]: Output should not be printed in this function directly. We should rather print the information in the command handler.

File: packages/sakuli-legacy/src/init-command/create-structure.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        #sakuli.autoHighlight=false
+        #sakuli.highlight.seconds=0.2
+        #sakuli.typeDelay=300
+        #sakuli.clickDelay=0.2
+        #sakuli.encryption.key=""
+        #sakuli.screenshot.onError=true
+        #sakuli.screenshot.dir=\${sakuli.log.folder}/_screenshots
+        `,
+        {encoding: 'utf8'});
+    console.log('Created file testsuite.properties');
```
Comment (tnobody) [inline packages/sakuli-legacy/src/init-command/create-structure.ts]: Output should not be printed in this function directly. We should rather print the information in the command handler.

File: packages/sakuli-legacy/src/init-command/create-structure.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+};
+
+const createTestsuiteSuite = (path: string) => {
+    writeFileSync(`${path}/testsuite.suite`, 'case1/check.js https://sakuli.io', {encoding: 'utf8'});
+    console.log('Created file testsuite.suite');
+};
+
+const createTestcaseInTestsuite = (path: string) => {
+    mkdirSync(`${path}/case1`, {recursive: true});
+    console.log('Created directory /case1');
```
Comment (tnobody) [inline packages/sakuli-legacy/src/init-command/create-structure.ts]: Output should not be printed in this function directly. We should rather print the information in the command handler.

File: packages/sakuli-legacy/src/init-command/create-structure.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    writeFileSync(`${path}/case1/check.js`, "", {encoding: 'utf8'});
+    console.log('Created file /case1/check.js');
```
Comment (tnobody) [inline packages/sakuli-legacy/src/init-command/create-structure.ts]: Output should not be printed in this function directly. We should rather print the information in the command handler.

File: packages/sakuli-legacy/src/init-command/create-structure.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        #sakuli.screenshot.onError=true
+        #sakuli.screenshot.dir=\${sakuli.log.folder}/_screenshots
+        `,
+        {encoding: 'utf8'});
+    console.log('Created file testsuite.properties');
+};
+
+const createTestsuiteSuite = (path: string) => {
+    writeFileSync(`${path}/testsuite.suite`, 'case1/check.js https://sakuli.io', {encoding: 'utf8'});
+    console.log('Created file testsuite.suite');
```
Comment (tnobody) [inline packages/sakuli-legacy/src/init-command/create-structure.ts]: Output should not be printed in this function directly. We should rather print the information in the command handler.

File: packages/sakuli-legacy/src/init-command/create-structure.ts [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        #sakuli.log.maxAge=14
+        #sakuli.log.folder=\${sakuli.testsuite.folder}/_logs
+        #sakuli.environment.similarity.default=0.99
+        #sakuli.autoHighlight=false
+        #sakuli.highlight.seconds=0.2
+        #sakuli.typeDelay=300
+        #sakuli.clickDelay=0.2
+        #sakuli.encryption.key=""
+        #sakuli.screenshot.onError=true
+        #sakuli.screenshot.dir=\${sakuli.log.folder}/_screenshots
```
Comment (tnobody) [inline packages/sakuli-legacy/src/init-command/create-structure.ts]: contents of generated files could be externalized into a "template" file or a single module which expose it as a constant.

File: packages/sakuli-legacy/src/init-command/create-structure.ts [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        main: "index.js",
+        scripts: {
+            test: `sakuli run ${suiteName}`
+        },
+        author: "",
+        license: "ISC",
+        dependencies: {
+            "@sakuli/cli": "^2.1.3",
+        },
+    };
```
Comment (tnobody) [inline packages/sakuli-legacy/src/init-command/create-structure.ts]: contents of generated files could be externalized into a "template" file or a single module which expose it as a constant.

**Comment by svettwer** [comment]
@tnobody it seems that github did not recognize your requested changes have been applied. Shall we dismiss the review?

One thought on this feature in general:
To execute `sakuli create project` sakuli has to be installed on the system or at least in a project containing a package.json with sakuli-cli. So this is not a tool to bootstrap the whole Sakuli project, right? You have to have a valid npm project with sakuli in it.

To put it differently: You cannot init a sakuli project without sakuli installed previously. :sweat_smile:

File: packages/sakuli-legacy/src/init-command/create-structure-files.ts [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+        version: "1.0.0",
+        description: "",
+        main: "index.js",
+        scripts: {
+            test: `sakuli run ${suiteName}`
+        },
+        author: "",
+        license: "ISC",
+        dependencies: {
+            "@sakuli/cli": "^2.1.3",
```
[the remark this answers is not in the document]
Comment (tnobody) [reply packages/sakuli-legacy/src/init-command/create-structure-files.ts]: Thank you for this hint. I think that is a very complex issue...
On the one hand we could discuss about generating this file at all (I'll explain my thoughts in another comment) on the other hand (when we do) we can face several issues:
- what version should we hardcore? When we hard code the 'current' version we will also publish this with the next release... But than it is one version behind
- We could always update the version quiet before the release but this is a fragile process

Best way (IMHO) would be: reading the `package.json` from this particular package (something like `join(__dirname, '..', 'package.json')` should access it, or read output from `sakuli --version`). So we can ensure that the installed version is the same as the global installed Sakuli version

**Comment by tnobody** [comment]
@svettwer
From what I see changes are made properly :)

You are right: The user have to install Sakuli globally first via `npm i -g @sakuli/cli` to have the command available on the system. Alternatively it __should__ also work when a user runs `npx sakuli create project ... --package` (even without installed Sakuli).
